lms/views/decorators/h_api.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+
+"""View decorators for working with the h API."""
+
+from __future__ import unicode_literals
+
+import json
+import requests
+
+from pyramid.httpexceptions import HTTPBadGateway
+from pyramid.httpexceptions import HTTPBadRequest
+from pyramid.httpexceptions import HTTPGatewayTimeout
+
+from lms import util
+from lms.util import MissingToolConsumerIntanceGUIDError
+from lms.util import MissingUserIDError
+
+
+def create_h_user(wrapped):  # noqa: MC0001
```

Comment:
I've ended up disabling a McCabe complexity linter warning here (`MC0001`) and also a pylint `too-many-branches` warning below. This warnings appeared at the very end of writing this PR, and did make me realise that this view decorator is too complex and doing too many things, but its taken a long time to get to this PR and I don't want to delay it any longer, so disabling the warnings for now.

Where I think I'd like to go with the factoring of this in the future is to add the services layer back into this app (it was removed by AJ) and have an `HAPIService` with methods for creating users etc with the h API, and another service (can't think of a good name for it yet) with methods for generating h user and group properties from the LTI params in the request. Feature flagging can go into a service too. So a lot of code'll be moved out of the view decorator and it'll end up looking something like:

```python
feature_flag_svc = request.find_service("feature_flag")
h_api_svc = request.find_service("h_api")
user_data_svc = request.find_service("user_data")

if not feature_flag_svc.is_enabled("auto_provisioning"):
    return wrapped(request, jwt)

try:
    h_api_svc.create_user(
        username=user_data_svc.username,  # user_data_svc has access to the request so can generate username based on request params
        display_name=user_data_svc.display_name,
        provider=user_data_svc.provider,
        provider_unique_id=user_data_svc.provider_unique_id,
    )
except LTIParamError as err:  # Raised by user_data_svc
    raise HTTPBadRequest(...)
except HTTPGatewayTimeout:  # Raised by h_api_svc
    raise
except HTTPBadGateway:  # Raised by h_api_svc
    raise

return wrapped(context, request)
```